### PR TITLE
Imviz: astrowidget API for center_on and offset_to

### PR DESCRIPTION
**Pull request by pllim**
Implement some of #631 and #632.

- [x] Make it work. Why is it zooming? See https://github.com/astropy/astrowidgets/pull/131/files#r585698007 (`delay_callback` is very important...)
- [x] See if `SkyCoord` implementation can be trivial. (Yes)
- [x] Add tests.
- [x] Update example notebook.

Out of scope: Unable to use this API on a tiled viewers; i.e., it only works on the first viewer that the app opened. See #643

Screenshot 2021-06-18 163551

### CODE REVIEW

File: jdaviz/configs/imviz/helper.py
Comment (pllim) [inline jdaviz/configs/imviz/helper.py:97]: @eteq , center/offset using sky coordinates isn't MVP, is it?

File: notebooks/ImvizExample.ipynb
Comment (pllim) [inline notebooks/ImvizExample.ipynb:5]: This isn't me. Jupyter did this after I upgraded. 🤷

File: jdaviz/configs/imviz/helper.py
Comment (pllim) [reply jdaviz/configs/imviz/helper.py:98]: Nevermind, I think I can fit this in.

File: jdaviz/configs/imviz/helper.py
Comment (pllim) [inline jdaviz/configs/imviz/helper.py:162]: Note: This was added in astropy/astropy#11635

File: notebooks/ImvizExample.ipynb
Comment (pllim) [inline notebooks/ImvizExample.ipynb:57]: For my sanity, I am replacing the example images with real HST/ACS data. Everyone loves 47 Tuc, right? Or is it just calibration team? 😉

File: notebooks/ImvizExample.ipynb
Comment (pllim) [inline notebooks/ImvizExample.ipynb:164]: Also for my sanity. Things are easier to see with 95th percentile.

File: jdaviz/configs/imviz/tests/test_astrowidgets_api.py
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_astrowidgets_api.py:11]: Note: This is basically the same test setup in #630 . When we get back to the other PR, will have to see how to consolidate this to avoid code duplication.

File: jdaviz/configs/imviz/helper.py
Comment (astrofrog) [inline jdaviz/configs/imviz/helper.py:109]: I think you should probably use image = viewer.reference_data as that is the one which is used to define the pixel space

File: jdaviz/configs/imviz/helper.py
Comment (astrofrog) [inline jdaviz/configs/imviz/helper.py:150]: as above

**Review by astrofrog — DISMISSED** [review]
Just a couple of comments but looks good otherwise!

File: jdaviz/configs/imviz/helper.py
Comment (pllim) [reply jdaviz/configs/imviz/helper.py:109]: Your suggestion does not seem to work: `AttributeError: 'ImvizImageView' object has no attribute 'reference_data'`

Comment (pllim) [reply jdaviz/configs/imviz/helper.py:109]: `viewer.state.reference_data` exists but it causes my test to fail:

```
        # Blink to the one without WCS
        self.viewer.blink_once()
    
        with pytest.raises(AttributeError, match='does not have a valid WCS'):
>           self.imviz.center_on(sky)
E           Failed: DID NOT RAISE <class 'AttributeError'>

jdaviz/configs/imviz/tests/test_astrowidgets_api.py:83: Failed
```

Comment (pllim) [reply jdaviz/configs/imviz/helper.py:109]: It appears `reference_data` does not keep track of which layer is actually visible to the user after blinking.

**Comment by rosteen** [comment]
Looks pretty good, the only thing I think might be good to add is some checking to see if the `center_on` input is within the bounds of the image. Inputting pixel coordinates beyond the bounds of the image centers there and just shows no data, which I guess is fine. Trying to `center_on` with sky coordinates that are outside the image, on the other hand, leads to a long stack trace:

```
---------------------------------------------------------------------------
NoConvergence                             Traceback (most recent call last)
<ipython-input-25-3978131487f0> in <module>
      3 
      4 sky = SkyCoord('01h24m07.33s -71d52m50.71s')
----> 5 imviz.center_on(sky)

~/projects/jdaviz/jdaviz/configs/imviz/helper.py in center_on(self, point)
    109             image = viewer.layers[i_top].layer
    110             if hasattr(image, 'coords') and isinstance(image.coords, BaseHighLevelWCS):
--> 111                 pix = image.coords.world_to_pixel(point)  # 0-indexed X, Y
    112             else:
    113                 raise AttributeError(f'{image.label} does not have a valid WCS')

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/astropy/wcs/wcsapi/high_level_api.py in world_to_pixel(self, *world_objects)
    230 
    231         # Finally we convert to pixel coordinates
--> 232         pixel = self.low_level_wcs.world_to_pixel_values(*world)
    233 
    234         return pixel

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/astropy/wcs/wcsapi/fitswcs.py in world_to_pixel_values(self, *world_arrays)
    315 
    316     def world_to_pixel_values(self, *world_arrays):
--> 317         pixel = self.all_world2pix(*world_arrays, 0)
    318         return pixel[0] if self.pixel_n_dim == 1 else tuple(pixel)
    319 

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/astropy/wcs/wcs.py in all_world2pix(self, tolerance, maxiter, adaptive, detect_divergence, quiet, *args, **kwargs)
   1884             raise ValueError("No basic WCS settings were created.")
   1885 
-> 1886         return self._array_converter(
   1887             lambda *args, **kwargs:
   1888             self._all_world2pix(

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/astropy/wcs/wcs.py in _array_converter(self, func, sky, ra_dec_order, *args)
   1334                     "a 1-D array for each axis, followed by an origin.")
   1335 
-> 1336             return _return_list_of_arrays(axes, origin)
   1337 
   1338         raise TypeError(

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/astropy/wcs/wcs.py in _return_list_of_arrays(axes, origin)
   1288             if ra_dec_order and sky == 'input':
   1289                 xy = self._denormalize_sky(xy)
-> 1290             output = func(xy, origin)
   1291             if ra_dec_order and sky == 'output':
   1292                 output = self._normalize_sky(output)

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/astropy/wcs/wcs.py in <lambda>(*args, **kwargs)
   1886         return self._array_converter(
   1887             lambda *args, **kwargs:
-> 1888             self._all_world2pix(
   1889                 *args, tolerance=tolerance, maxiter=maxiter,
   1890                 adaptive=adaptive, detect_divergence=detect_divergence,

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/astropy/wcs/wcs.py in _all_world2pix(self, world, origin, tolerance, maxiter, adaptive, detect_divergence, quiet)
   1868                     slow_conv=ind, divergent=None)
   1869             else:
-> 1870                 raise NoConvergence(
   1871                     "'WCS.all_world2pix' failed to "
   1872                     "converge to the requested accuracy.\n"

NoConvergence: 'WCS.all_world2pix' failed to converge to the requested accuracy.
After 2 iterations, the solution is diverging at least for one input point.
```

Perhaps it would be useful to check whether the sky coordinates input are within the image bounds, and if not raise a more informative/concise error that you can't center on sky coordinates outside the image.

**Comment by pllim** [comment]
@rosteen , I think @mcara said this error only happens if image is distorted...? Which image did you use? Thanks for smoking this out!

**Comment by rosteen** [comment]
That's with `acs_47tuc_1` from the Imviz example notebook.

**Comment by pllim** [comment]
Ah, yes. FLC is not corrected for distortion. 😬

**Comment by pllim** [comment]
@rosteen , is this better?

Screenshot 2021-06-25 121305

**Comment by rosteen** [comment]
I like capturing it that way much better than the long stack trace in the notebook. Approving, since making this work with more viewers needs to happen after some other work.

**Review by rosteen — APPROVED** [review]

**Comment by pllim** [comment]
@astrofrog 's suggestion did not work for me. Given that the current code works, should we just merge?

**Comment by rosteen** [comment]
Let's see if we can be good and get a second approval, since you said you have other things you can work on today.

**Comment by ojustino** [comment]
I'm trying it out but have to toggle between tasks. I will give an update before the end of the day.

**Review by ojustino — COMMENTED** [review]
I reviewed using some images of Cygnus A from astroquery. `center_on` and `offset_to` both worked for me.

When I attempted to center on a location far outside the images I didn't see a stack trace or a pop up. The viewer went white and I was unable to get back to the images without restarting.

It might be a separate issue, but none of the layer manipulation settings (stretch, cmap, percentile, etc.) worked for me whether I set them through code or the viewer. I tried notebook and Lab in my usual `jdaviz` environment and a freshly installed one.

**Comment by pllim** [comment]
Thanks!

> The viewer went white and I was unable to get back to the images without restarting

Try commenting out the cell that ignore warnings. Do you see any warnings that might be relevant?

> layer manipulation settings

If you have this same problem with the code from `main`, then it is a separate issue and you should file a bug report.

**Comment by rosteen** [comment]
> It might be a separate issue, but none of the layer manipulation settings (stretch, cmap, percentile, etc.) worked for me whether I set them through code or the viewer. I tried notebook and Lab in my usual `jdaviz` environment and a freshly installed one.

Any chance you had loaded multiple images, and were adjusting the layer setting for a layer that wasn't on top?

**Comment by ojustino** [comment]
@rosteen That was indeed my mistake for the layer manipulation issue, thanks.

**Comment by ojustino** [comment]
@pllim There are no warnings that fire immediately after running the `center_on` with an off-image point. Only after I mouse over the (whited-out) viewer do I see this print repeatedly in the error log:
```
TypeError                                 Traceback (most recent call last)
~/opt/anaconda3/envs/jdaviz-dev5/lib/python3.8/site-packages/glue_jupyter/bqplot/common/viewer.py in _on_mouse_interaction(self, interaction, data, buffers)
    151     def _on_mouse_interaction(self, interaction, data, buffers):
    152         for callback in self._event_callbacks:
--> 153             callback(data)
    154 
    155     @debounced(delay_seconds=0.5, method=True)

~/repositories/jdaviz/jdaviz/configs/imviz/plugins/viewers.py in on_mouse_or_key_event(self, data)
     67             fmt = 'x={0:0' + str(maxsize) + '.1f} y={1:0' + str(maxsize) + '.1f}'
     68 
---> 69             self.label_mouseover.pixel = (fmt.format(x, y))
     70 
     71             if hasattr(image, 'coords') and isinstance(image.coords, BaseHighLevelWCS):

TypeError: unsupported format string passed to NoneType.__format__
```

**Comment by pllim** [comment]
@ojustino , it is unrelated but since you found it, might as well try to fix. Can you please help me debug by adding `print(x, y)` right before `self.label_mouseover.pixel = (fmt.format(x, y))` in your installed copy of `viewers.py` and show me their values?

**Comment by ojustino** [comment]
@plim I get floats until I center on an offending location. Then it's `None None` repeatedly.

**Comment by pllim** [comment]
Okay, looks like we just need to special case for `None` in the coordinates info bar. I'll push out a fix shortly. Thanks for checking!

**Comment by pllim** [comment]
@ojustino , could you please install the updated branch  here and try again? Thanks! 🙏

**Comment by ojustino** [comment]
Just did. No more error printouts, but the viewer still whites out and trying to recenter on an in-image location doesn't have a visible effect.

**Comment by pllim** [comment]
Added one more check. Is this a satisfactory solution to the error you encounter, @ojustino ?

Screenshot 2021-06-25 162821

**Comment by pllim** [comment]
The CI failure smells like upstream problem in Actions:

```
  Error: fatal: unable to access 'https://github.com/spacetelescope/jdaviz/': The requested URL returned error: 500
  The process '/usr/bin/git' failed with exit code 128
```

**Comment by ojustino** [comment]
@pllim With this commit, `center_on` triggers the new dropdown message and blocks execution if I try SkyCoords or pixels that are off-screen but do exist "on-image."

It is not blocking SkyCoords that are off-screen and off-image, and that case still causes the white out and does not trigger the dropdown message.

**Comment by pllim** [comment]
@ojustino , at this point... please point me to that image you are using, and provide me with the calls that (1) block execution, and (2) white out.

Thanks for the very thorough checks!

**Review by ibusko — APPROVED** [review]
I was able to run the demo notebook and exercise all cells that have the centering and offseting operations. The image seemed to behave in the way we expect.

**Comment by ojustino** [comment]
Perhaps the problems are with my environment, then? In any case, these are my images:
```
cygA1 = download_file('https://hla.stsci.edu/cgi-bin/getdata.cgi?dataset=hst_05368_01_wfpc2_f555w_pc_drz.fits', cache=True)
cygA2 = download_file('https://hla.stsci.edu/cgi-bin/getdata.cgi?dataset=hst_05368_01_wfpc2_f555w_wf_drz.fits', cache=True)
```

This is the offending code:
```
from astropy.coordinates import SkyCoord
sky = SkyCoord('09h59m26.02s 40d44m17.92s')
```

I also tried the notebook's original 47 Tuc images and get the same unintended blocking behavior in the "off-screen, on-image" case, though the "off-screen, off-image" case is blocked as intended. I should also note that anything "on-screen and on-image" has been working fine.

**Comment by pllim** [comment]
What does "off-screen, on-image" mean?

**Comment by ojustino** [comment]
A set of coordinates that has data in the image but is not currently visible in the viewer. An example would be if I'm zoomed in at one corner and want to center on a location on the opposite side of the image.

**Comment by pllim** [comment]
> A set of coordinates that has data in the image but is not currently visible in the viewer.

Oh... In this case, perhaps changing x/y_min/max of the layer does not actually load the data that is currently not displayed. If you know how to force data loading, please let me know. Otherwise, here is my proposal:

1. Let me know if 7039327 needs to be reverted or not.
2. Once (1) is resolved, we merge and open new issue about the "off-screen, on-image" use case. This way, we can wait for SME, while I can also continue onto #699 while we wait till they become available.

**Comment by ojustino** [comment]
That's an answer I don't know.

1. I think 7039327 is a good idea for avoiding the whiteout scenario, so I'm fine with a merge, especially since other developers didn't encounter any issues.
2. I'm fine with migrating this problem into a new issue.

**Comment by pllim** [comment]
@astrofrog , if you feel strongly about using `i_top` to grab the top visible layer, we can handle it in a follow-up PR. Thanks!

**Comment by pllim** [comment]
Thank you for the very thorough reviews, everyone!

@ojustino , I opened #704 as a follow-up issue. Please correct me if I got anything wrong there. Thanks again!

**Comment by eteq** [comment]
Post-hoc review: this looks great, except for one thing: I think it should be `offset_by` isntead of `offset_to`!  Will create a follow-on issue to capture that.

**Comment by pllim** [comment]
@eteq , I think that ship has sailed? 

https://github.com/astropy/astrowidgets/blob/ace5abb820c030122add475ecf106114aaa68e1b/astrowidgets/core.py#L346

https://github.com/eteq/nb-astroimage-api/blob/ee6f0a30673b11c0c68192c6c2fd1dfa47ceb23c/minimum_viable_imagewidget.py#L53

**Pull request merged**
